sycl/doc/extensions/template.rst

```diff
@@ -0,0 +1,653 @@
+Instead, just describe the semantics and APIs of the extension itself.
+We will figure out later how to change the core SYCL specification if the
+extension is adopted.
+
+Avoid unnecessary duplication of the SYCL specification
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Avoid unnecessary duplication of APIs or text from the core SYCL specification.
+For example, if your extension adds a new member function to an existing class,
+there is no need to duplicate the existing member functions in your
+specification.
+Duplication like this can be troublesome if something in the core SYCL
+specification changes later.
+
+Use ISO C++ "words of power"
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+When specifying a function, use paragraphs with the following titles:
+
+* *Constraints:* This paragraph can be used whenever the function is templated
+  or when the function is a member of a templated class.
+  The constraints describe the SFINAE constraints or ``requires`` constraints
+  of the template parameters.
+
+* *Preconditions:* This paragraph tell the conditions that the application must
+  obey when calling the function.
+  If the application violates a precondition, the behavior is undefined.
+
+* *Effects:* Tells what the function does.
+
+* *Returns:* Tells what value the function returns.
+
+* *Throws:* Tells what exceptions the function is required to throw and the
+  circumstances under which they are thrown.
+
+* *Remarks:* Tells other information about the function if it is not covered by
+  the previous paragraphs.
+
+All these terms are also used in the ISO C++ specification, and we use them in
+the SYCL specification with the same meaning they have in C++.
+
+Often the *Constraints*, *Preconditions*, or *Throws* paragraphs will have
+several items.
+When this happens, use a bulleted list.
+
+Only include a paragraphs if its term is needed to specify the function.
+Simply omit paragraphs whose terms are unnecessary.
+
+Normative vs. non-normative text
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+A specification is usually a mixture of normative and non-normative text.
+Normative text is the formal description of the extension, while non-normative
+text is intended to provide general context and clarity.
+A good specification should describe all of its features using normative text.
+In other words, it should be possible for someone to implement your extension
+just by reading the normative text.
+
+The main section titled "Specification" (and its subsections) are the normative
+part of the specification document.
+All other sections (e.g. "Overview", "Examples", etc.) are non-normative.
+In addition, examples are non-normative even if they appears inside the
+"Specification" section.
+This applies to example code and also to sentences that describe an example.
+(Such sentences often start with "For example, ...".)
+Note that the code synopses are not examples, and therefore they are normative.
```

Comment:
Is it worth noting here (or nearby) that there's a precedent in normative synopses for:
- Unspecified types, etc, and the format we use to describe those; and
- "Exposition only" functions and members, which may be included to improve readability but should not be implemented